src/parsers.ts

```diff
@@ -1,25 +1,61 @@
-import { Dictionary } from './basic';
-import { IValidationResult } from './validations';
+import { IDiagnostic } from './validations';
 
 export type SourceMapParser<T = any> = (value: string) => IParserResult<T>;
 
-export type ObjPath = Array<string | number>;
+export type DocumentUri = string;
+export type Segment = string | number;
+export type JSONPath = Segment[];
+
+export interface IParserResult<T = any> {
+  data: T;
+  diagnostics: IDiagnostic[];
+
+  // One of these, not sure? jsonc uses offset, language server spec uses line + character :/
+  // whatever we choose we need to be able to support for all of our parsers (json, yaml, markdown)
+  getJsonPathForPosition(position: IPosition): JSONPath | undefined;
+
+  // ILocation comes from the language server specification
+  getLocationForJsonPath(path: JSONPath): ILocation | undefined;
+}
 
 export interface IPosition {
-  /** line number (should start at 1) */
-  line: number;
-  column?: number;
+  /**
+   * Line position in a document (zero-based).
+   */
+  readonly line: number;
+
+  /**
+   * Character offset on a line in a document (zero-based). Assuming that the line is
+   * represented as a string, the `character` value represents the gap between the
+   * `character` and `character + 1`.
+   *
+   * If the character value is greater than the line length it defaults back to the
+   * line length.
+   */
+  readonly character: number;
 }
 
-export interface IPathPosition {
-  start: IPosition;
-  end?: IPosition;
+/**
+ * A range represents an ordered pair of two positions.
+ * It is guaranteed that start isBeforeOrEqual end.
+ */
+export interface IRange {
+  /**
+   * The start position. It is before or equal to end.
+   */
+  readonly start: IPosition;
+
+  /**
+   * The end position. It is after or equal to start.
+   */
+  readonly end: IPosition;
 }
 
-export type IParserResultPointers = Dictionary<IPathPosition, string>;
+export interface ILocation {
+  uri: DocumentUri;
```

Comment:
Should we make this optional?

Reply:
@marbemac Yeah, I never actually provide any value at the moment, so making it optional makes sense.